### ISSUE TITLE
Enhancement: [prefer-promise-reject-errors] option to allow 'rethrow' of signal reasons or caught values

### ISSUE THREAD

**Issue by turbocrime** [typescript-eslint/typescript-eslint#11095]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/prefer-promise-reject-errors/

### Description

When performing more complex async operations, a developer may call the `reject` executor of a `Promise` to propagate an `AbortSignal` reason, a value obtained by a `catch` block, or the reason parameter of a `.then(_, onrejected)` or `.catch()` callback.

in this case, it's often desirable to *reject with the exact reason*, to accurately propagate failures without changing them.

but, `prefer-promise-reject-errors` may complain unless these values are consistently verified.

### Fail

```TypeScript
// the rule must be disabled to rethrow D: this should be unnecessary
function rejectOnSignal(signal: AbortSignal) {
  return new Promise<never>((_, reject) => {
    // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors -- rethrow
    signal.addEventListener('abort', () => reject(signal.reason));
  });
}
```

### Pass

```TypeScript
// yay the rule allows rethrow :D it's not necessary to disable it
function rejectOnSignal(signal: AbortSignal) {
  return new Promise<never>((_, reject) => {
    signal.addEventListener('abort', () => reject(signal.reason));
  });
}
```

### Additional Info

the example is rather simple and contrived, but surprisingly common - an event listener is the most certain and reliable way to immediately react to signal activation.

the behavior may be achieved a couple other ways, which the rule doesn't complain about, but i think that's rather an additional issue (the rule is inconsistent) rather than an acceptable workaround.

for example, this is permitted without complaint, but assignment to an event handler idl attribute is undesirable because it clobbers existing handlers. possibly this should be another issue

```ts
function rejectOnSignal(signal: AbortSignal) {
  return new Promise<never>((_, reject) => {
    signal.onabort = reject;
  });
}
```

and this is permitted without complaint, but it appears to mostly be a failure to understand the `reject` of `Promise.withResolvers` as a promise executor. possibly this should be another issue

```ts
function rejectOnSignal(signal: AbortSignal) {
  const { promise, reject } = Promise.withResolvers<never>();
  signal.addEventListener('abort', () => reject(signal.reason));
  return promise;
}
```

**Comment by kirkwaiblinger** [comment]
Closely related: #10376, #10717

**Comment by kirkwaiblinger** [comment]
Hi @turbocrime! 

Reading more closely, I think there's several separately actionable things discussed here: 

1. Support `Promise.withResolvers()` in the rule.
1. Support `abortSignal` in the rule
   - I propose we defer this / consider this a duplicate to the outcome of #10717 
1. Support `somePromise.catch((e) => { Promise.reject(e); });` (and its corollaries)
   - This has meaningful precedent in favor with #10376 being accepted.

Would you say that's accurate? If so, I'm thinking we constrain this issue to be about 3. and I'd invite you to file a new issue to discuss 1. WDYT?

**Comment by JoshuaKGoldberg** [comment]
👋 @turbocrime just checking in, do you have any thoughts on the suggestions?

**Comment by JoshuaKGoldberg** [comment]
Closing out as it's been a few months without interaction. If someone has a compelling case for 1 or 3, please do post here (if it's not yet locked) or file a new issue. Cheers!

**Comment by alexeyr-ci2** [comment]
I do not know what a compelling case is, but I found this issue because I wanted 3. Or for that matter
```
catch (e) {
   return Promise.reject(e);
}
```
Currently the best approximation is to enable both `allowThrowingAny` (for `.catch(e => ...)`) and `allowThrowingUnknown` (for `catch (e) { ... }`), which allow much wrong code that this option wouldn't (as described in https://github.com/typescript-eslint/typescript-eslint/issues/10376#issuecomment-2498329402).
